css/css-display/display-contents-accessibility.html

```diff
@@ -0,0 +1,321 @@
+    and its pseudo-elements, such as ::before and ::after pseudo-elements, which are generated before/after the element’s children as normal)."
+
+    - Per CSS spec, setting "display: contents" must not alter an element's semantics (https://www.w3.org/TR/css-display-3/#valdef-display-contents):
+    "As only the box tree is affected, any semantics based on the document tree, such as selector-matching, event handling, and
+    property inheritance, are not affected."
+
+    - Categorizing behavior of "display: contents" for HTML elements (https://www.w3.org/TR/css-display-3/#unbox-html):
+        (1) Compute the same as "display: none":
+            <br>
+            <wbr>
+            <meter?
+            <progress>
+            <canvas>
+            <embed>
+            <object>
+            <audio>
+            <iframe>
+            <img>
+            <video>
+            <frame>
+            <frameset>
+            <input>
+            <textarea>
+            <select>
+
+        (2) Not rendered, but behave as intended with "display: contents":
+            <legend>
+
+        (3) No special behavior, behave as intended with "display: contents":
+            <button>
+            <details>
+            <fieldset>
+
+        (4) Behave as intended with "display: contents":
+            All other HTML elements.
+
+    - As of December 2023, the current state of cross-browser, automated web accessibility testing is limited to retrieving the following
+    accessibility object properties with planned future enhancements (https://github.com/WICG/aom/issues/203):
+        - computedRole
+        - computedLabel
+
+    Thus, the following has been currently identified as testable when "display: contents" is supplied:
+        - HTML elements retain their semantic role (such as <h#>, <th>, <td>, <button>, <svg>)
+        - HTML elements and their accessible name calculation is unaffected (https://w3c.github.io/accname/#computation-steps)
+
+    The following is not currently WPT testable:
+        - Focusability of controls (https://github.com/mozilla/standards-positions/issues/772)
+
+        - Reading order such as table cells
+
+        - Expected keyboard operability such as links remaining operable with Enter/Return
+
+        - Hidden elements as a result of "display: contents"
+
+        - <table> structure such as accessibility row or column count
+    -->
+
+    <h1>Testing accessibility of display: contents</h1>
+
+    <p>Verifies accessibility behavior of the <a href="https://www.w3.org/TR/css-display-3/#box-generation">contents</a> keyword for the CSS display property</p>
+
+    <h2>Accname</h2>
+
+    <button aria-labelledby="b2" style="display: contents;" data-expectedlabel="hidden label" data-testname="button with display:contents is labelled via referenced hidden span" class="ex-label">
```

Comment:
Seems like it might be a good idea to test both role and label expectations on all test elements (except generics and other prohibited roles). Since we don't yet have direct WebDriver access to `isIgnored` or equivalent platform API, We're trying to confirm `display:contents` doesn't cause unexpected output (or more likely lack of output) in most of these, and failing affectations to either of the testable properties would constitute unexpected behavior.

Of course I'm very open to disagreement on this point, if there's something I'm overlooking.

```suggestion
    <button aria-labelledby="b2" style="display: contents;" data-expectedrole="button" data-expectedlabel="hidden label" data-testname="button with display:contents is labelled via referenced hidden span" class="ex-role ex-label">
```

Reply:
Maybe there are places where other implementation differences result in different roles (layout lists, for example) so we could note that with a comment and specifically not test it here.

Reply:
@cookiecrook Good idea, updated all applicable tests to test bothrole/label: https://github.com/web-platform-tests/wpt/pull/43740/commits/1808f7bc51320a52326153543d3c598ea8b9c2c0.

> Maybe there are places where other implementation differences result in different roles (layout lists, for example) so we could note that with a comment and specifically not test it here.

Are there any other examples? Happy to document them.

Reply:
I think I misunderstood; when I add both `class="ex-role ex-label"`, this results in many duplicate test names. It doesn't appear that they are treated as a single test...

Reply:
Oh yeah... I think you'd need a local function to run them both and modify the test name with a prefix or suffix...

Reply:
Or you could add a ~`verifyRoleAndLabelBySelector()` to the utils file that passes an optional, modified test name string to the existing `verify…BySelector` methods.

Reply:
Agreed, please take a look: 08f76c0bd1d96b6d8067b18ea523832dd40a1670.